x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/UriUtils.java

```diff
@@ -42,42 +56,211 @@ public static URI parseURI(String connectionString, URI defaultURI) {
+            StringBuilder sb = new StringBuilder();
+            if (uri.getScheme() != null) {
+                sb.append(uri.getScheme());
+                sb.append("://");
+            }
+            if (uri.getRawUserInfo() != null) {
+                sb.append("\0".repeat(uri.getRawUserInfo().length()));
+                if (uri.getHost() != null) {
+                    sb.append('@');
+                }
+            }
+            if (uri.getHost() != null) {
+                sb.append(uri.getHost());
+            }
+            if (hasPort || uri.getPort() > 0) {
+                sb.append(':');
+            }
+            if (uri.getPort() > 0) {
+                sb.append(uri.getPort());
+            }
+            if (uri.getRawPath() != null) {
+                sb.append(uri.getRawPath());
+            }
+            if (uri.getQuery() != null) {
+                sb.append('?');
+                // redact with the null character; this will later allow safe reinsertion of any character removed from the URI to make
+                // it parsable
+                sb.append(redactCredentialsInRawUriQuery(uri.getRawQuery(), '\0'));
+            }
+            if (uri.getRawFragment() != null) {
+                sb.append('#');
+                sb.append(uri.getRawFragment());
+            }
+
+            // reinsert any removed character back into the URI: if the reinsertion should be made between null characters, replace its
+            // value with a null character, since it's part of the credential value
+            Collections.reverse(faults);
+            for (Map.Entry<Integer, Character> e : faults) {
+                int idx = e.getKey();
+                if (idx >= sb.length()) {
+                    sb.append(e.getValue());
+                } else {
+                    sb.insert(idx,
+                        (sb.charAt(idx) == '\0' && (idx + 1 >= sb.length() || sb.charAt(idx + 1) == '\0')) ? '\0' : e.getValue());
+                }
+            }
+
+            StringBuilder ret = new StringBuilder();
+            sb.chars().forEach(x -> ret.append(x == '\0' ? REDACTION_CHAR : (char) x));
+
+            return ret.toString();
+        }
+
+        private static String redactCredentialsInURLString(String urlString) {
+            List<Map.Entry<Integer, Character>> faults = new ArrayList<>();
+
+            boolean hasPort = false;
+            for (StringBuilder sb = new StringBuilder(urlString); sb.length() > 0; ) {
+                try {
+                    // parse as URL; ex. `http://ho~st` parses as URI, but with unparsable authority
+                    URI uri = new URI(sb.toString()).parseServerAuthority();
+                    return editURI(uri, faults, hasPort);
+                } catch (URISyntaxException use) {
+                    int idx = use.getIndex();
+                    if (idx < 0 || idx >= sb.length()) {
```

Comment:
When can the `idx` be `> sb.length()`? And why the equality is not a valid use case to be redacted?

Reply:
> When can the `idx` be > `sb.length()`?

_Probably_ never, but not sure what URISyntaxException can provide as "valid" value there.

> And why the equality is not a valid use case to be redacted?

In this case the index would fall outside the input string length; for instance parsing "http://" as URI will return `Expected authority at index 7: http://`, for which the equality holds.